tests/test_templates_parameters.py

```diff
@@ -0,0 +1,341 @@
+
+    def test_pars_cvqnnlayers_normal_range(self, seed):
+        """Confirm that pennylane.templates.utils.parameters_cvqnnlayers_normal() samples from the right
+        distributions."""
+        p = parameters_cvqnnlayers_normal(n_layers=2, n_wires=1000, mean=-1, std=0.2, seed=seed)
+        p_av = np.array([np.mean(p_) for p_ in p])
+        p_std = np.array([np.std(p_) for p_ in p])
+        target_av = np.array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1])
+        target_std = np.array([0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2])
+        assert np.allclose(p_av, target_av, atol=0.13)
+        assert np.allclose(p_std, target_std, atol=0.13)
+
+
+class TestParsInterferometer:
+    """Tests the pennylane.templates.parameters method for an interferometer."""
+
+    def test_pars_interferometer_uniform_dimensions(self, n_subsystems):
+        """Confirm that pennylane.templates.utils.parameters_interferometer_uniform()
+         returns an array with the right dimensions."""
+        a = (n_subsystems, )
+        b = (n_subsystems * (n_subsystems - 1) // 2, )
+        p = parameters_interferometer_uniform(n_wires=n_subsystems, seed=0)
+        dims = [p_.shape for p_ in p]
+        assert dims == [b, b, a]
+
+    def test_pars_interferometer_uniform_range(self, seed):
+        """Confirm that pennylane.templates.utils.parameters_interferometer_uniform() samples from the right
+        distributions."""
+        p = parameters_interferometer_uniform(n_wires=1000, uniform_min=-2, uniform_max=1, seed=seed)
+        p_av = np.array([np.mean(p_) for p_ in p])
+        p_std = np.array([np.std(p_) for p_ in p])
+        target_av = np.array([-0.5, -0.5, -0.5])
+        target_std = np.array([0.86, 0.86, 0.86])
+        assert np.allclose(p_av, target_av, atol=0.13)
+        assert np.allclose(p_std, target_std, atol=0.13)
+
+    def test_pars_interferometer_normal_dimensions(self, n_subsystems):
+        """Confirm that pennylane.templates.utils.parameters_interferometer_normal()
+         returns an array with the right dimensions."""
+        a = (n_subsystems, )
+        b = (n_subsystems * (n_subsystems - 1) // 2, )
+        p = parameters_interferometer_normal(n_wires=n_subsystems, seed=0)
+        dims = [p_.shape for p_ in p]
+        assert dims == [b, b, a]
+
+    def test_pars_interferometer_normal_range(self, seed):
+        """Confirm that pennylane.templates.utils.parameters_interferometer_normal() samples from the right
+        distributions."""
+        p = parameters_interferometer_normal(n_wires=1000, mean=-1, std=0.2, seed=seed)
+        p_av = np.array([np.mean(p_) for p_ in p])
+        p_std = np.array([np.std(p_) for p_ in p])
+        target_av = np.array([-1, -1, -1])
+        target_std = np.array([0.2, 0.2, 0.2])
+        assert np.allclose(p_av, target_av, atol=0.13)
+        assert np.allclose(p_std, target_std, atol=0.13)
+
+
+class TestParsStronglyEntangling:
```

Comment:
There is one serious theoretical issue with this. The standard error, the deviation of a set of means computed from samples from the true mean, is given by the std deviation of the initial distribution divided by sqrt(sample_size). To get a standard error of the order of 1e-4, one needs of the order of 100,000,000 samples...The random function runs out of memory, and even with a workaround this slows down the test a lot.

A second issue is that one has to create multiple arrays, rather than simply increasing the number of wires, since the number of IF parameters explodes with the wires. But that's of course solvable...

I suggest a compromise of a special "sampling tolerance" of 1e-2, which I try to reach in all tests.

Reply:
That's actually pretty difficult enough, I just checked. It increases the test time from a fraction of a second to up to 10 seconds per test where this is necessary...

Reply:
:+1:
Instead of using an absolute tolerance in the assert statements, would a better approach be to use a relative tolerance? i.e. `atol=0, rtol=something)`?

Another alternative is to make the absolute tolerance be dependent on the std deviation, i.e., `atol=3/np.sqrt(sample_size)`.

Reply:
> That's actually pretty difficult enough, I just checked. It increases the test time from a fraction of a second to up to 10 seconds per test where this is necessary...

I would favour faster tests over higher accuracy in cases where sampling occurs.

Reply:
Both might help to take into account the factor of the standard deviation in the standard error, but it's still only a linear scaling factor if I understand correctly? It's certainly more theory-driven than just setting the tolerance to a higher threshold!

Let's hear what Nathan says?

Reply:
Let's think about this a bit more. There has to be some way to unit test the basic functionality (that samples are returned according to the proper distribution) without excessive overhead. How does the official `numpy.random` get tested itself? Maybe we just want to confirm that the samples are the same as `numpy.random` for a fixed seed, and not worry about statistics?

Reply:
> Maybe we just want to confirm that the samples are the same as numpy.random for a fixed seed, and not worry about statistics?

thumb

Reply:
I'm not convonced: For the CVQNN the test would literally have to reproduce sampling 11 arrays, and simply copy most of the function code into the test...

A quick online search shows that on the related topic of "testing random number generators", testing the expected statistical properties of mean and standard deviation is indeed a common option! 

Is hand-engineering the tolerance to a theoretically predicted value really so wrong? Even if the tolerance is rather high?

Reply:
After a little bit of research, including looking into how [`numpy.random`](https://github.com/numpy/numpy/blob/master/numpy/random/tests/test_random.py) itself is tested I've come to the following thoughts:

i) we *can* unit test simple things, like a function returns an argument with the correct shape (we already do this)
ii) we can also unit test deterministic edge cases, e.g., sampling from normal distribution with `std=0`, or a uniform distribution with `low==high` (we should do this)
iii) beyond this, unit testing is simply not a great paradigm for testing of statistical quantities and distributions. Even `numpy.random` does not seem to test the statistical properties of their samples; instead, each function is tested with a single random seed and a single target output. IMO, these are more tests that verify the seed functionality doesn't break when changes are made. 

Based on this, I make the following proposal:
- we add tests for the simple edge cases where these distributions collapse to fixed points
- we add simple tests that these functions return the same result as the `numpy.random` versions of these functions for some fixed seed (even in the CVQNN params example, I think it is worth it) -- this will at least ensure we are in agreement with `numpy` if we change things in the future
- we don't do anything further

@josh146 @mariaschuld Thoughts?

Reply:
Sounds very reasonable. For uniform one could also test that no sample lies outside the interval. 

Ok if I implement these on Monday?

Reply:
That sounds great @co9olguy, I'm in favour. Does this new 'policy' affect any existing stochastic tests?

Reply:
@mariaschuld No problem
@josh146 Let's consider this question separately for other projects. What is reasonable here might not apply elsewhere

Reply:
New tests written!

Reply:
Testing edge cases (uniform [a,a]/normal std=0), seeds, and interval for uniform.

Also added new tests to do the same for layerS functions.